Astro Cloud IDE documentation (#1371)
astronomer/docs: astro/cloud-ide/quickstart.md

---
sidebar_label: Quickstart
title: Astro Cloud IDE quickstart
id: quickstart
---

:::caution

<!-- id to make it easier to remove: cloud-ide-preview-banner -->

> The Cloud IDE is currently in _Public Preview_. This means that it is available to all Astro customers, but is still undergoing heavy development and features are subject to change. Creating projects in the Cloud IDE does not impact running tasks on existing Deployments. If you have any feedback, please submit it to the [Astro Cloud IDE product portal](https://portal.productboard.com/75k8qmuqjacnrrnef446fggj).

Reviewer: ```suggestion
The Cloud IDE is currently in _Public Preview_ and it is available to all Astro customers. It is still in development and features and functionality are subject to change. Creating projects in the Cloud IDE does not affect running tasks on existing Deployments. If you have feedback on the Cloud IDE you'd like to share, you can submit it on the [Astro Cloud IDE product portal](https://portal.productboard.com/75k8qmuqjacnrrnef446fggj).
```

:::

> This quickstart walks you through creating and running your first project through the Astro Cloud IDE.

Reviewer: ```suggestion
Use this quickstart to create and run your first project with the Cloud IDE.
```

## Prerequisites 

> To complete this quickstart, you need:

Reviewer: ```suggestion
To complete this quickstart, you need:
```
Remove.

> - Workspace Editor permissions to an [Astro Workspace](manage-workspaces.md).

Reviewer: ```suggestion
- Workspace Editor permissions in an [Astro Workspace](manage-workspaces.md).
```

- Optional. A database hosted in one of the following services:
    - GCP BigQuery
    - Postgres (hosted)
    - Snowflake
    - AWS S3
    - Redshift 

> If you don't provide a database, you can still complete the tutorial. However, you will not be able to test any SQL in your pipeline.

Reviewer: ```suggestion
If you don't provide a database, you can still complete the quickstart. However, you won't be able to test SQL code in your pipeline.
```

## Step 1: Log in and create a project

> The Cloud IDE is accessible through the Cloud UI and is available to all Astro customers.

Reviewer: ```suggestion
The Cloud IDE is available to all Astro customers and can be accessed in the Cloud UI.
```

1. Log in to the Cloud UI and select a Workspace. 
2. Click **Cloud IDE** in the left sidebar. If you are the first person in your Workspace to use the Astro Cloud IDE, the **My Projects** page will be empty.

![Cloud IDE](/img/cloud-ide/project-list.png)

> 3. Click **+ Project**. Give your new project a name and a description, then click **Create**.

Reviewer: ```suggestion
3. Click **Project**, enter a name and a description for the project, and then click **Create**.
```

> After you create your project, the Astro Cloud IDE opens your project home page. This page contains all menus for configuring your your project:

Reviewer: ```suggestion
After you create your project, the Cloud IDE opens your project home page with the following tabs:
```

- The **Pipelines** tab stores all of the Python and SQL code that your project executes.
- The **Connections** tab stores Airflow connections for connecting your project to external services.

> - The **Variables** tab stores Airflow variables for use in your pipeline code.

Reviewer: ```suggestion
- The **Variables** tab stores Airflow variables used in your pipeline code.
```

- The **Requirements** tab stores the required Python and OS-level dependencies for running your pipelines.

## Step 2: Create a pipeline

> In the **Pipelines** tab, click **+ Pipeline**. Give your pipeline a name and a description, then click **Create**.

Reviewer: ```suggestion
1. Click the **Pipelines** tab and then click **Pipeline**. 

2. Enter a name and a description for the pipeline and then click **Create**.
```

When you first run your project, your pipeline is built into a single DAG with the name you provide. Because of this, pipeline names must be unique within their project. They must also be a Python identifier, so they can't contain spaces or special characters. 

After clicking **Create**, the IDE opens the pipeline editor. This is where you'll write your pipeline code.

## Step 3: Create a Python cell 

Just like tasks are the building blocks of DAGs, cells are the building blocks of pipelines. Cells can be created with either Python or SQL code. For this tutorial, write a Python cell named `hello_world`.

> 1. Click **Create Cell**, then click **Python**. A new cell named `cell_1` appears on the page.

Reviewer: ```suggestion
1. In the **Pipeline** list, click the name of the pipeline you created in step 2.
2.  Click **Add Cell** and select **Python**. A new cell named `cell_1` appears.
```

2. Click the cell's name and rename the cell `hello_world`.
3. Click the editing window where it says `# Write your code here…`, then replace that line with the following code:
   
   ```python
   return "Hello, world!"
   ```

Your pipeline editor should look like the following:

![Hello world cell in the pipeline editor](/img/cloud-ide/hello-world.png)

## Step 4: Run your cell 

In the `hello_world` cell, click **Run** to execute a single run of your cell.

![Python Cell Logs](/img/cloud-ide/python-cell-logs.png)

> When you run a cell, The Astro Cloud IDE creates a temporary DAG that executes only individual cell. Astro runs this DAG in your cloud and sends information about the cell run directly to the Astro IDE.

Reviewer: ```suggestion
When you run a cell, the Cloud IDE creates a temporary DAG that executes only an individual cell. Astro runs this DAG in your cloud and sends information about the cell run directly to the Cloud IDE.
```

Reviewer: This didn't work for me.

> When you run a cell, The Astro Cloud IDE creates a temporary DAG that executes only individual cell. Astro runs this DAG in your cloud and sends information about the cell run directly to the Astro IDE.

Reviewer 2: @jlaneve What actually happens here? Took a wild guess, but I have no idea where the DAG runs (Deployment? Venv? Control plane? Data plane?)

> The **Logs** tab contains all logs generated by the cell run, including Airflow logs and Python errors. The **Results** tab contains the contents of your Python console. If you click **Results** you should see the result of your successful cell run. 

Reviewer: ```suggestion
The **Logs** tab contains all logs generated by the cell run, including Airflow logs and Python errors. The **Results** tab contains the contents of your Python console. Click **Results** to view the result of your successful cell run. 
```

## Step 5: Create a database connection

To create a SQL cell and execute SQL, create a database to run your SQL queries against. 

> 1. Click **Environment**, then click **Connections**. 

Reviewer: ```suggestion
1. Click the **Connections** tab and then click **Connection**. 
```

    ![Configure Connection](/img/cloud-ide/new-connection.png)

> 2. Click **NEW CONNECTION**.  

Reviewer: ```suggestion
2. Click **NEW CONNECTION**.  
```
Remove

3. Choose one of the available connection types and configure all required values for the connection. Click **More options** to configure optional values for the connection.

  :::info

  SQL cell query results are stored as temporary tables in the database you're querying. The schema field in your connection object determines where these end up. Astronomer recommends setting this field to `TMP_ASTRO` for easy cleanup.

  :::

> 4. Optional. Click **Test Connection**. The Astro Cloud IDE runs a quick connection test and displays a message based on whether the test was a success. Regardless of whether the test succeeds, you can still create the connection.

Reviewer: ```suggestion
4. Optional. Click **Test Connection**. The Astro Cloud IDE runs a quick connection test and returns a status message. You can still create the connection if the test is unsuccessful.
```

> 4. Optional. Click **Test Connection**. The Astro Cloud IDE runs a quick connection test and displays a message based on whether the test was a success. Regardless of whether the test succeeds, you can still create the connection.

Reviewer 2: @jlaneve I did some photoshop here for the sake of the guide, but I actually couldn't get the connection to work or even an error message when testing my connection. Is this a known issue, and if so should we be documenting this aspect of the IDE yet?

5. Click **Create Connection**. You new connection appears in the **Connections** tab both in the pipeline editor and on your project homepage. You can use this connection with any future pipelines you create in this project. 

## Step 6: Create a SQL cell

You can now write and run SQL cells with your database connection.

> 1. Click **Create Cell**, then click **SQL**. A new cell appears on the page.

Reviewer: ```suggestion
1. In the **Pipeline** list, click the name of the pipeline you created in step 2.

2.  Click **Add Cell** and select **SQL**. A new cell named `cell_1` appears.
```

> 2. Click the cell's name and rename the cell `hello_sql`.

Reviewer: ```suggestion
2. Click the cell name and rename it `hello_sql`.
```

> 3. In the dropdown menu next to the cell's name, select your database connection.

Reviewer: ```suggestion
3. In the **Select Connection** list, select the connection you created in step 5.
```

> 4. Click the editing window where it says `-- Write your SQL query here`, then replace that line with the following code:

Reviewer: ```suggestion
4. Replace `-- Write your SQL query here` with the following code:
```

    ```sql
    SELECT 1 AS hello_world;
    ```

  :::tip

>   You can also quick add a SQL cell with a given connection by clicking on the **+** button from the **Connections** tab in the **Environment** menu.

Reviewer: ```suggestion
  You can also add a SQL cell to a connection by clicking the **+** button from the **Connections** tab in the **Environment** menu.
```

  :::

> 5. Optional. Click **Run** to test the SQL query. The results of your query appear in the **Results** tab.

Reviewer: ```suggestion
5. Optional. Click **Run** to test the SQL query. The results of your query appear in the **Results** tab.
```

I didn't see a **Results** tab.

## Step 7: Create dependencies between cells

You now have a Python cell and a SQL cell, but there's currently no logic to determine which task runs first in your DAG. You can create dependencies for these cells directly in the Astro Cloud IDE. Start by making `hello_sql` dependent on `hello_world` completing.

> 1. In the `hello_sql` cell, click the **Dependencies** icon.

Reviewer: ```suggestion
1. In the `hello_sql` cell, click **Dependencies** and then select `hello_world`.
```

2. Click `hello_world`. 
   
    ![Configure a dependency](/img/cloud-ide/create-dependency.png)   

> 3. To confirm that the dependency was established, click **Pipeline**. The **Pipeline** menu visualizes the dependencies between your cells.

Reviewer: ```suggestion
3. To confirm that the dependency was established, click **Pipeline**. The **Pipeline** menu visualizes the dependencies between your cells.
```
Is this step necessary? The Pipeline view is the default view. If you keep this step, you'll probably want to add an image to show the location of the Pipeline icon.

    ![Dependency graph in the Pipelines menu](/img/cloud-ide/configured-dependency.png)   

> ## Step 7: Generate dependencies in code

Reviewer: ```suggestion
## Step 7: Generate dependencies in code
```
This is the second Step 7. Confirm the numbering sequence is correct.

One of the most powerful features of the Astro Cloud IDE is that it can automatically detect data dependencies in your cell code and restructure your pipeline based on those dependencies. This works for both Python and SQL cells.

To create a potential dependency to a Python cell, the upstream Python cell must end with a `return` statement. This means that you can create a downstream dependency from `hello_world`.

1. Create a new Python cell named `data_dependency`.
2. Add the following code to the cell:

    ```sh
    my_string = hello_world
    return my_string
    ``` 

    You can pass any value from a `return` statement into a downstream Python cell by calling the name of the upstream Python cell.

> 3. Open **Pipeline** menu to see that your dependency graph has updated:

Reviewer: ```suggestion
3. Click **Pipeline** to confirm that your dependency graph was updated:
```

    ![New dependency graph](/img/cloud-ide/data-dependency.png) 

> You can generate dependencies between any two types of cells. See the following topics for more information about generating different types of dependencies. 

Reviewer: ```suggestion
You can generate dependencies between any two types of cells. 
```

The following Python cell is downstream of a SQL cell named `my_sql_cell`. 

```python
df = my_sql_cell # my_sql_cell is a SQL cell which gets converted to a pandas DataFrame by default
df['col_a'] = df['col_a'] + 1
return df
```

### Use the results of a Python cell in a SQL cell

You can make a SQL cell downstream dependent of a Python cell as long as the Python cell returns a pandas DataFrame. To do so, reference the name of the Python cell in your SQL cell using double curly braces, also known as jinja templating. Pandas DataFrames are automatically converted to SQL tables for use in SQL cells.

The following SQL cell is downstream of a Python cell named `my_dataframe`. 

```sql
select * from {{my_dataframe}} -- my_dataframe is a Python cell
where col_a > 10
```

## Step 8: Run your pipeline 

Now that you've completed your pipeline, you can run it from beginning to end. To do so, click **Run** in the top right corner of your pipeline editing window. Cells are executed in order based on their dependencies. During the run, the **Pipeline** page shows which cells have been executed and which are still pending.

![Run Pipeline](/img/cloud-ide/run-pipeline.png)

> ## Step 9 Schedule your pipeline

Reviewer: ```suggestion
## Step 9: Schedule your pipeline
```

> Once you've verified that your pipeline works properly, you can schedule it to run on a regular basis. 

Reviewer: ```suggestion
After you've verified that your pipeline is working, you can schedule it to run regularly. 
```

> To set your pipeline's schedule, click **Schedule**. 

Reviewer: ```suggestion
1. To set your pipeline's schedule, click **Schedule**. 
```

![Schedule Pipeline](/img/cloud-ide/schedule-menu.png)

> The Cloud IDE currently supports scheduling with cron. You can either manually type in a cron string or click **EDIT** to open the cron builder, which is a simple UI for setting a cron schedule. After you make your selections and close out the cron builder, the Astro Cloud IDE loads a newly generated cron schedule.

Reviewer: ```suggestion
The Cloud IDE currently supports scheduling with cron. You can manually enter a cron string, or you can click **EDIT** to open the cron builder, which is a simple UI for setting a cron schedule. After you make your selections and close the cron builder, the Astro Cloud IDE loads a newly generated cron schedule.
```

> When you finish configuring your schedule, click **Update Settings** to save your changes. 

Reviewer: ```suggestion
Click **Update Settings** to save your changes. 
```
